### PR TITLE
EP-251: Page Viewed (manage_pledge)

### PR DESCRIPTION
**Pull request by sunday-okpoluaefe**
# 📲 What

Android: Page Viewed (manage_pledge)

# 🤔 Why

Segment Implementation

# 🛠 How

- Create method ``` trackManagePledgePageViewed ``` 
 ``` 
  /**
     * Sends data associated with page view event to the client.
     *
     * @param projectData: The selected project data.
     * @param checkoutData: checkout data
     * @param addOns: associated addons
     */
    fun trackManagePledgePageViewed(projectData: ProjectData, checkoutData: CheckoutData, addOns: List<Reward>?) {
        val props: HashMap<String, Any> = hashMapOf(CONTEXT_PAGE.contextName to MANAGE_PLEDGE.contextName)
        props.putAll(AnalyticEventsUtils.projectProperties(projectData.project(), client.loggedInUser()))
        props.putAll(AnalyticEventsUtils.checkoutProperties(checkoutData, projectData.project(), addOns))
        props.putAll(AnalyticEventsUtils.refTagProperties(projectData.refTagFromIntent(), projectData.refTagFromCookie()))
        client.track(PAGE_VIEWED.eventName, props)
    }

```

- Called ``` trackManagePledgePageViewed ``` from ``` BackingFragmentViewModel ```

```
 this.isExpanded
                .filter { it }
                .compose(combineLatestPair(backing))
                .map { it.second }
                .compose<Pair<Backing, ProjectData>>(combineLatestPair(projectDataInput))
                .compose(bindToLifecycle())
                .subscribe {
                    this.lake.trackManagePledgePageViewed(it.first, it.second)
                }
```

- Added segment properties unit test 
```
  @Test
    fun testManagePledgePageViewed() {
        val project = ProjectFactory.backedProject()
            .toBuilder()
            .id(4)
            .category(CategoryFactory.ceramicsCategory())
            .commentsCount(3)
            .creator(creator())
            .location(LocationFactory.unitedStates())
            .updatesCount(5)
            .build()

        val addOn1 = RewardFactory.addOn()
        val addOn2 = RewardFactory.addOnMultiple()

        val backing = BackingFactory.backing().toBuilder()
            .project(project)
            .addOns(listOf(addOn1, addOn2))
            .bonusAmount(35.0)
            .shippingAmount(20f)
            .location(LocationFactory.germany())
            .locationId(LocationFactory.germany().id())
            .build()

        val creator = creator()
        val client = client(creator)
        client.eventNames.subscribe(this.segmentTrack)
        client.eventProperties.subscribe(this.propertiesTest)
        val segment = AnalyticEvents(listOf(client))

        val projectData = ProjectDataFactory.project(project, RefTag.discovery(), RefTag.recommended())

        segment.trackManagePledgePageViewed(backing = backing, projectData = projectData)

        assertSessionProperties(creator)
        assertProjectProperties(project)
        assertContextProperties()

        val expectedProperties = this.propertiesTest.value

        // - we test asserting this properties all methods in SharedFunctions.kt
        assertEquals(10.0, expectedProperties["checkout_amount"])
        assertEquals("credit_card", expectedProperties["checkout_payment_type"])
        assertEquals(30.0, expectedProperties["checkout_amount_total_usd"])
        assertEquals(20.0, expectedProperties["checkout_shipping_amount_usd"])
        assertEquals(5, expectedProperties["checkout_add_ons_count_total"])
        assertEquals(2, expectedProperties["checkout_add_ons_count_unique"])
        assertEquals(100.0, expectedProperties["checkout_add_ons_minimum_usd"])
        assertEquals(35.0, expectedProperties["checkout_bonus_amount_usd"])
        assertEquals(MANAGE_PLEDGE.contextName, expectedProperties[CONTEXT_PAGE.contextName])

        this.segmentTrack.assertValue(EventName.PAGE_VIEWED.eventName)
    }

```

- Added unit tests to page viewed events 
```
  /*
    * Page viewed event is expected when fragment is visible --> isExpanded = true
    */
    @Test
    fun testManagePledgePageViewed_withExpanded_True() {
        val project = ProjectFactory.project()
        setUpEnvironment(environment())

        this.vm.inputs.configureWith(ProjectDataFactory.project(project))
        this.vm.isExpanded(true)

        this.segmentTrack.assertValue(EventName.PAGE_VIEWED.eventName)
    }
```
    /*
     * No event is expected when fragment is not visible --> isExpanded = false
     */
    @Test
    fun testManagePledgePageViewed_withExpanded_False() {
        val project = ProjectFactory.project()
        setUpEnvironment(environment())

        this.vm.inputs.configureWith(ProjectDataFactory.project(project))

        this.vm.isExpanded(false)

        this.segmentTrack.assertNoValues()
    }

```

# 👀 See

<img width="331" alt="Screenshot 2021-03-29 at 10 53 08" src="https://user-images.githubusercontent.com/63934292/112821102-31c1de00-907e-11eb-9019-f88410dcc483.png">


| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Go to profile, from the list of backed projects, select any.
- You should the new event registered on the segment console

<img width="736" alt="Screenshot 2021-03-29 at 10 38 11" src="https://user-images.githubusercontent.com/63934292/112821219-5322ca00-907e-11eb-8956-5704249917db.png">


# Story 📖

https://kickstarter.atlassian.net/browse/EP-251

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/activities/BackingActivity.java [hunk 1]
```diff
@@ -22,7 +22,10 @@ public void onCreate(final @Nullable Bundle savedInstanceState) {
 
     this.viewModel.outputs.showBackingFragment()
             .compose(bindToLifecycle())
-            .subscribe(this::startBackingFragment);
+            .subscribe(it -> {
+              this.startBackingFragment(it);
+              this.backingFragment().pageViewed();
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/BackingActivity.java:27]: No need to trigger the page Viewd event here, it should be fired on the BackingFragment

File: app/src/main/java/com/kickstarter/ui/activities/ProjectActivity.kt [hunk 2]
```diff
@@ -290,6 +290,11 @@ class ProjectActivity :
             .observeOn(AndroidSchedulers.mainThread())
             .subscribe { this.startVideoActivity(it) }
 
+        this.viewModel.outputs.managePledgePageViewed()
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/activities/ProjectActivity.kt:293]: No logic related to events should be placed on an Activity or Fragment

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt [hunk 3]
```diff
@@ -247,6 +247,9 @@ interface ProjectViewModel {
 
         /** Emits when we should update the [com.kickstarter.ui.fragments.BackingFragment] and [com.kickstarter.ui.fragments.RewardsFragment].  */
         fun updateFragments(): Observable<ProjectData>
+
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt:250]: No new logic should be added on ProjectViewModel just for triggering the new event on Manage pledge screen

File: app/src/main/java/com/kickstarter/ui/fragments/BackingFragment.kt [hunk 4]
```diff
@@ -303,8 +303,14 @@ class BackingFragment : BaseFragment<BackingFragmentViewModel.ViewModel>() {
 
     fun configureWith(projectData: ProjectData) {
         this.viewModel.inputs.configureWith(projectData)
+        this.activity?.let {
```
Comment (Arkariang) [inline app/src/main/java/com/kickstarter/ui/fragments/BackingFragment.kt:306]: Delete this part, after the merge of this PR this is no longer needed #1186

File: app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt (hunk 3 above)
Comment (sunday-okpoluaefe) [reply app/src/main/java/com/kickstarter/viewmodels/ProjectViewModel.kt]: done

**Review by Arkariang — APPROVED** [review]

**Pull request merged**